### PR TITLE
myFT engagement stored in redis and available via myft-engagement route

### PR DESCRIPTION
**Pull request by laurieboyes**
e.g. If we go here: `/v2/next/myft-engagement/f0bb6f11-c034-4792-a1c0-3647825363a0`

We get something like this:
```
{
	follow_events_last_30_days: "4",
	save_events_last_30_days: "0",
	preference_on_events_last_30_days: "5",
	push_notifications_clicked_last_30_days: "0",
	email_clicked_events_last_30_days: "0",
	feature_usage_events_last_30_days: "0",
	push_notifications_shown_last_7_days: "0",
	email_open_events_last_7_days: "0",
	following: "2",
	opted_into_daily_email: "true",
	opted_into_weekly_email: "false",
	is_myft_user: "true"
}
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -58,6 +58,7 @@ run:
 	nbt run --local
 
 run-local:
+	export PORT=5454 \
```
Comment (phamann) [inline Makefile]: Why?

File: jobs/engagement.js [hunk 2]
```diff
@@ -1,71 +1,108 @@
...
+const nodeIdToday = new NodeIdToday();
+
+const tomorrow = require('../server/v2/lib/tomorrow-id');
+const userQuery = `MATCH (user:User) WHERE user.uuid = {userId}`;
+const timeQueryStrings = {
+	month: `MATCH (tomorrow:Day) WHERE id(tomorrow) = {tomorrowId}
+						MATCH (tomorrow)<-[:NEXT*..30]-(lastMonth:Day)`,
+	week: `MATCH (tomorrow:Day) WHERE id(tomorrow) = {tomorrowId}
+						MATCH (tomorrow)<-[:NEXT*..7]-(lastWeek:Day)`,
+	now: ``
```
Comment (phamann) [inline jobs/engagement.js]: Should this be empty?

File: jobs/engagement.js [hunk 3]
```diff
@@ -1,71 +1,108 @@
...
+
 worker.setup({
 	name: (process.env.IS_TEST_APP === 'true') ? 'ft-next-myft-api-test' : 'ft-next-myft-api'
 }).then(() => {
 
 	new worker.CronJob({
 		cronTime: '00 4 * * *', // every day at 4am
-		onTick: refreshCachedUserEngagement
+		onTick: refreshUserEngagement
 	});
```
Comment (phamann) [inline jobs/engagement.js]: Shouldn't this config have `start: true`?

File: jobs/lib/all-users-query.js [hunk 4]
```diff
@@ -3,13 +3,16 @@
 const _ = require('lodash');
 const directly = require('directly');
 const tap = require('./tap');
-const FlagClearer = require('./flag-clearer');
+const LabelClearer = require('./label-clearer');
 const db = require('../utils/db');
+const redisClient = require('../../server/utils/redis-client');
+
+// todo this is now tightly coupled to engagement and needs some major rethinking
```
Comment (phamann) [inline jobs/lib/all-users-query.js]: :see_no_evil:

File: jobs/engagement.js (hunk 2 above)
Comment (laurieboyes) [inline jobs/engagement.js]: Yep but it probably wants a comment. We don't need to get a timetree node for this query so we leave it blank, for the sake consistnency in using the generateQuery function

File: Makefile (hunk 1 above)
Comment (laurieboyes) [inline Makefile]: It's pretty arbitrary, but for this: https://github.com/Financial-Times/next-myft-page/pull/579/files

File: jobs/lib/all-users-query.js (hunk 4 above)
Comment (laurieboyes) [inline jobs/lib/all-users-query.js]: Ha, yes. Some refactoring work to do here

File: jobs/engagement.js (hunk 3 above)
Comment (laurieboyes) [inline jobs/engagement.js]: Taken care of by n-worker: https://github.com/Financial-Times/n-worker/blob/389770ee36baf2c9f39eb16612c7cdac9b5b7237/index.js#L80

File: jobs/lib/label-clearer.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
+'use strict';
+
+const db = require('../utils/db');
+const labelPrefixWhiteList = [
+	'Flagged',
+	'Locked'
+];
+
+class LabelClearer {
```
Comment (phamann) [inline jobs/lib/label-clearer.js]: `LabelRemover` ?

File: jobs/lib/label-clearer.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	constructor (label) {
+
+		if(!labelPrefixWhiteList.reduce((prev, curr) => prev || label.startsWith(curr), false)) {
+			throw new Error('LabelClearer will not accept non-whitelisted label ' + label);
+		}
+
```
Comment (phamann) [inline jobs/lib/label-clearer.js]: I've also used an internal `this.count` value previously to keep track an log how many items were cleared.

File: server/utils/redis-client.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+const redisGreen = url.parse(process.env.REDIS_URL);
+
+const redisClient = redis.createClient({
+	port: redisGreen.port,
+	host: redisGreen.hostname,
+	enable_offline_queue: false // fail fast when redis is down
+});
+
+redisClient.on("error", err => {
+	console.log('Redis Error', err);
```
Comment (phamann) [inline server/utils/redis-client.js]: Maybe import `Metrics` and log this to Graphite.

File: server/v2/controllers/next/myft-engagement.js [hunk 8]
```diff
@@ -1,104 +1,23 @@
...
-			}
-
-			const data = results
-							.reduce((a,b) => a.concat(b))
-							.reduce((a, b) => {
-								let key = Object.keys(b)[0];
-								a[key] = b[key];
-								return a;
-							}, {});
+	redisClient.hgetall(userId, (err, userEng) => {
```
Comment (phamann) [inline server/v2/controllers/next/myft-engagement.js]: Consider having a more specific key than the users `userd`. Maybe something like `userId:engagement`

**Comment by phamann** [comment]
Awesome! Looks good to me other than comments. Excited to get this working! :+1:

File: jobs/engagement.js (hunk 2 above)
Comment (laurieboyes) [inline jobs/engagement.js]: Actually scratch that, I'll get rid of this smelly stuff

**Pull request merged**